modules/index-service/src/main/java/org/opencastproject/index/service/message/WorkflowMessageReceiverImpl.java

```diff
@@ -71,110 +49,69 @@ public WorkflowMessageReceiverImpl() {
-          logger.warn("Workflow instance mediapackage {} not found for deletion", eventId);
-        } catch (SearchIndexException e) {
-          logger.error("Error deleting the Workflow instance entry {} from the search index: {}", eventId,
-                  ExceptionUtils.getStackTrace(e));
-        }
-        return;
-      case AddDefinition:
-        // TODO: Update the index with it as soon as the definition are part of it
-        return;
-      case DeleteDefinition:
-        // TODO: Update the index with it as soon as the definition are part of it
-        return;
+        deleteWorkflowInstance(workflowItem);
+        break;
       default:
         throw new IllegalArgumentException("Unhandled type of WorkflowItem");
     }
   }
 
-  /** OSGi DI. */
-  public void setWorkspace(Workspace workspace) {
-    this.workspace = workspace;
-  }
-
-  /** OSGi DI. */
-  public void setAuthorizationService(AuthorizationService authorizationService) {
-    this.authorizationService = authorizationService;
+  private void deleteWorkflowInstance(WorkflowItem workflowItem) {
+    final String organization = getSecurityService().getOrganization().getId();
+    final User user = getSecurityService().getUser();
+    final String eventId = workflowItem.getId();
+    logger.debug("Received Delete Workflow instance Entry {}", eventId);
+
+    // Remove the Workflow instance entry from the search index
+    try {
+      getSearchIndex().deleteWorkflow(organization, user, eventId, workflowItem.getWorkflowInstanceId());
+      logger.debug("Workflow instance media package {} removed from search index", eventId);
+    } catch (NotFoundException e) {
+      logger.warn("Workflow instance media package {} not found for deletion", eventId);
+    } catch (SearchIndexException e) {
+      logger.error("Error deleting the Workflow instance entry {} from the search index", eventId, e);
+    }
   }
 
-  /** OSGi callback for acl services. */
-  public void setAclServiceFactory(AclServiceFactory aclServiceFactory) {
-    this.aclServiceFactory = aclServiceFactory;
+  private void updateWorkflowInstance(WorkflowItem workflowItem) {
+    logger.debug("Received Update Workflow instance Entry for index {}", getSearchIndex().getIndexName());
+    final String organization = getSecurityService().getOrganization().getId();
+    final User user = getSecurityService().getUser();
+    final String eventId = workflowItem.getId();
+    final MediaPackage mediaPackage = workflowItem.getMediaPackage();
+
+    // Load or create the corresponding recording event
+    try {
+      Event event = getOrCreateEvent(eventId, organization, user, getSearchIndex());
+      event.setCreator(user.getName());
+      event.setWorkflowId(workflowItem.getWorkflowInstanceId());
+      event.setWorkflowDefinitionId(workflowItem.getWorkflowDefinitionId());
+      event.setWorkflowState(workflowItem.getState());
+      event.setAccessPolicy(workflowItem.getAccessControlListJSON());
```

Comment:
The original source code did not update the ACL and Dublincore if the workflow instance is in state.isTerminated(). That seems reasonable as the workflow instance is not the data source for the event (the "archive" is).

To my understanding, the idea of creating and updating an event here is a workaround for the poor design idea to create an event as a side-effect of workflow execution during ingest - rather than creating the event at ingest and then start the workflow for this event which would make all this code totally obsolete as the workflow itself has neither metadata catalogs nor ACLs. The only thing this code would need to update is the event status as that one indeed does depend on the workflow instance...

... anyway, that is way out of scope of this PR, so back to this pull request:
Omitting the test !state.isTerminated() seems a step into the wrong direction to me and also seems to involve additional work load in case the workflow is in state.isTerminated().

Is there a specific reason you want to omit that test?

Reply:
>  the workflow instance is not the data source for the event (the "archive" is).

I'd like that to be true, but it isn't necessarily. There are some edge cases… It would also not  be sufficient to only update the workflow status. At least not right now. We can discuss that tomorrow if you need more information.

Now, the `isTerminated()` was actually a quite recent addition (commit e26c446953) as a workaround for the issue that data are requested in this message handler and the likelihood of this failing (remember that this is executed asynchronously) was very high compared to other message handling, were it was still a very real possibility.

Now, with the data already included in the message, there wasn't a real reason for this exception any longer.